ppsspp: add SDL and headless

SamLukeYes opened the pull request:
Change the default ppsspp package to SDL front end, as it supports
Vulkan, and is allowed to build the headless binary together.

###### Description of changes

- New packages
  - `ppsspp-sdl`
  - `ppsspp-sdl-wayland`

- Packages changed
  - `ppsspp` now uses SDL front end, includes the headless binary, and supports Vulkan
  - The original `ppsspp` with Qt front end is renamed to `ppsspp-qt`
  - Add a desktop entry for the packages above
 
- Additional notes
  - Qt front end doesn't support Vulkan yet. See https://github.com/hrydgard/ppsspp/issues/11628
  - SDL front end has trouble with Vulkan on Wayland, so Vulkan support is disabled by default in `ppsspp-sdl-wayland`. See https://github.com/hrydgard/ppsspp/issues/13845

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

AndersonTorres commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1493,7 +1493,24 @@ with pkgs;
+  };
+
+  ppsspp-sdl = callPackage ../applications/emulators/ppsspp {
+    pname = "ppsspp-sdl";
```
`pname` should not be a "visible parameter" to all-packages.nix.

Instead you can programatically change it inside the derivation, like we do with cmake:

https://github.com/NixOS/nixpkgs/blob/a50594822c73de196d148991eb110f6484fd123b/pkgs/development/tools/build-managers/cmake/default.nix#L35

AndersonTorres commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1493,7 +1493,24 @@ with pkgs;
+  ppsspp-sdl-wayland = callPackage ../applications/emulators/ppsspp {
+    pname = "ppsspp-sdl-wayland";
+    forceWayland = true;
+    enableVulkan = false; # https://github.com/hrydgard/ppsspp/issues/13845
```
It is preferrable to put such a documentation inside the expression.

AndersonTorres commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1493,7 +1493,24 @@ with pkgs;
+    enableVulkan = false; # https://github.com/hrydgard/ppsspp/issues/13845
+  };
+
+  ppsspp = callPackage ../applications/emulators/ppsspp {
```
Put this expression first.

AndersonTorres commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1493,7 +1493,24 @@ with pkgs;
+    enableVulkan = false; # https://github.com/hrydgard/ppsspp/issues/11628
+  };
+
+  ppsspp-sdl = callPackage ../applications/emulators/ppsspp {
```
After the rebase, write this and the other expressions in the form `ppsspp-sdl = ppsspp.override { . . . };`.

AndersonTorres commented on pkgs/applications/emulators/ppsspp/default.nix:
```diff
@@ -1,21 +1,46 @@
 , zlib
+
+, pname ? "ppsspp"
+
```
Get rid of these blank lines.

AndersonTorres commented on pkgs/applications/emulators/ppsspp/default.nix:
```diff
@@ -1,72 +1,128 @@
+    "-DUSING_QT_UI=${if enableQt then "ON" else "OFF"}"
   ];
 
+  desktopItem = makeDesktopItem {
```
```suggestion
  desktopItems = [(makeDesktopItem {
```

AndersonTorres commented on pkgs/applications/emulators/ppsspp/default.nix:
```diff
@@ -1,72 +1,128 @@
+    icon = "ppsspp";
+    comment = "Play PSP games on your computer";
+    categories = [ "Game" "Emulator" ];
+  };
```
```suggestion
  })];
```

AndersonTorres commented on pkgs/applications/emulators/ppsspp/default.nix:
```diff
@@ -1,72 +1,128 @@
+  '' + "\n" + lib.optionalString enableHeadless ''
+    install -Dm555 PPSSPPHeadless $out/bin/ppsspp-headless
+  '' + ''
+    install -Dm444 ${finalAttrs.desktopItem}/share/applications/ppsspp.desktop $out/share/applications
```
```suggestion
```

AndersonTorres commented on pkgs/applications/emulators/ppsspp/default.nix:
```diff
@@ -1,72 +1,128 @@
 
-  nativeBuildInputs = [ cmake pkg-config python3 ];
+  nativeBuildInputs = [
+    cmake
```
```suggestion
    cmake
    copyDesktopItems
```

AndersonTorres commented on pkgs/applications/emulators/ppsspp/default.nix:
```diff
@@ -1,72 +1,128 @@
+, makeWrapper
 , fetchFromGitHub
 , SDL2
 , cmake
```
```suggestion
, cmake
, copyDesktopItems
```

AndersonTorres commented on pkgs/applications/emulators/ppsspp/default.nix:
```diff
@@ -1,72 +1,128 @@
-{ mkDerivation
+{ lib
+, stdenv
+, makeDesktopItem
+, makeWrapper
```
```suggestion
```

AndersonTorres commented on pkgs/applications/emulators/ppsspp/default.nix:
```diff
@@ -1,72 +1,128 @@
-, lib
+, libffi
 , libzip
 , pkg-config
```
```suggestion
, makeDesktopItem
, makeWrapper
, pkg-config
```

AndersonTorres commented on pkgs/applications/emulators/ppsspp/default.nix:
```diff
@@ -1,72 +1,128 @@
+, wayland
+, wrapQtAppsHook ? null
 , zlib
+
```
```suggestion
```

AndersonTorres commented on pkgs/applications/emulators/ppsspp/default.nix:
```diff
@@ -1,72 +1,128 @@
+
+# Currently only enabled on SDL X11 builds
+, enableVulkan ? true
+
```
```suggestion
```

AndersonTorres commented on pkgs/applications/emulators/ppsspp/default.nix:
```diff
@@ -1,72 +1,128 @@
+
+# Headless build can only be enabled with SDL
+, enableHeadless ? true
+
```
```suggestion
```

AndersonTorres commented on pkgs/applications/emulators/ppsspp/default.nix:
```diff
@@ -1,72 +1,128 @@
+  # experimental, see https://github.com/hrydgard/ppsspp/issues/13845
+  vulkanWayland = enableVulkan && forceWayland;
+in
+
```
```suggestion
```

jian-lin commented on pkgs/applications/emulators/ppsspp/default.nix:
```diff
@@ -1,72 +1,128 @@
+    "-DUSING_QT_UI=${if enableQt then "ON" else "OFF"}"
   ];
 
+  desktopItem = makeDesktopItem {
```
Is it better to upstream this dekstop item?

AndersonTorres replied:
> Is it better to upstream this dekstop item?

This is not always desirable for many projects. Desktop items are XGD specific IIRC.

jian-lin reviewed (changes requested):
The relations between newly added parameters (`enableVulkan`, `enableHeadless` and `forceWayland`) are a bit confusing, could you try to:
- remove uneeded ones, e.g. removing `enableHeadless` and enabling it if possible
- and add `assert` to make sure these arguments are valid?

SamLukeYes committed:
ppsspp: add SDL and headless

Change the default ppsspp package to SDL frontend, as it supports
vulkan, and is allowed to build the headless binary together.

AndersonTorres commented on pkgs/applications/emulators/ppsspp/default.nix:
```diff
@@ -3,76 +3,119 @@
+      mkdir -p $out/share/{applications,ppsspp}
+    '' + (if enableQt then ''
+      install -Dm555 PPSSPPQt $out/bin/ppsspp
+      wrapProgram $out/bin/ppsspp \
```
What are the parameters of this wrapping?

SamLukeYes replied:
In case `enableVulkan` is overridden to be `true`, it would be wrapped with `--prefix LD_LIBRARY_PATH : ${vulkanPath}`. This should allow the Qt front end to detect Vulkan, though not able to use it for now.

AndersonTorres replied:
Ugly as hell.
